compiler/prelude/types.js

```diff
@@ -792,3 +863,234 @@ var $assertType = (value, type, returnTuple) => {
     }
     return returnTuple ? [value, true] : value;
 };
+
+const $isSigned = (typ) => {
+    switch (typ.kind) {
+        case $kindInt:
+        case $kindInt8:
+        case $kindInt16:
+        case $kindInt32:
+        case $kindInt64:
+            return true;
+        default:
+            return false;
+    }
+}
+
+/**
+ * Truncate a JavaScript number `n` according to precision of the Go type `typ`
+ * it is supposed to represent.
+ */
+const $truncateNumber = (n, typ) => {
+    switch (typ.kind) {
+        case $kindInt8:
+            return n << 24 >> 24;
+        case $kindUint8:
+            return n << 24 >>> 24;
+        case $kindInt16:
+            return n << 16 >> 16;
+        case $kindUint16:
+            return n << 16 >>> 16;
+        case $kindInt32:
+        case $kindInt:
+            return n << 0 >> 0;
+        case $kindUint32:
+        case $kindUint:
+        case $kindUintptr:
+            return n << 0 >>> 0;
+        case $kindFloat32:
+            return $fround(n);
+        case $kindFloat64:
+            return n;
+        default:
+            $panic(new $String("invalid kind: " + kind));
+    }
+}
+
+/**
+ * Trivial type conversion function, which only accepts destination type identical to the src
+ * type.
+ * 
+ * For wrapped types, src value must be wrapped, and the return value will be unwrapped.
+ */
+const $convertIdentity = (src, dstType) => {
+    const srcType = src.constructor;
+    if (srcType === dstType) {
+        // Same type, no conversion necessary.
+        return srcType.wrapped ? src.$val : src;
+    }
+    throw new Error(`unsupported conversion from ${srcType.string} to ${dstType.string}`);
+};
+
+/**
+ * Conversion to int64 and uint64 variants.
+ *
+ * dstType.kind must be either $kindInt64 or $kindUint64. For wrapped types, src
+ * value must be wrapped. The returned value is an object instantiated by the
+ * `dstType` constructor.
+ */
+const $convertToInt64 = (src, dstType) => {
+    const srcType = src.constructor;
+    if (srcType === dstType) {
+        return src.$val;
+    }
+
+    switch (srcType.kind) {
+        case $kindInt64:
+        case $kindUint64:
+            return new dstType(src.$val.$high, src.$val.$low);
+        case $kindUintptr:
+            // GopherJS quirk: a uintprt value may be an object converted to
```

Comment:
typo? Or is this a term I'm unfamiliar with?

```suggestion
            // GopherJS quirk: a uintptr value may be an object converted to
```

Reply:
Done.